plutus-core/plutus-core/src/PlutusCore/Flat.hs

```diff
@@ -334,25 +337,34 @@ instance Flat Special
 -- See Note [Index (Word64) (de)serialized through Natural]
 instance Flat Index where
     -- encode from word64 to natural
-    encode = encode @Natural . fromIntegral
+    encode =
+#if WORD_SIZE_IN_BITS == 64
+        encode . wordToNatural . fromIntegral @Index @Word
+#else
+        encode @Natural . fromIntegral
+#endif
     -- decode from natural to word64
     decode = do
-        n <- decode
-        case naturalToWord64Maybe n of
-            Just w  -> pure $ Index w
-            Nothing -> fail $ "Index outside representable range: " ++ show n
+        n <- decode @Natural
+#if WORD_SIZE_IN_BITS == 64
```

Comment:
I'd rather put this inside `naturalToWord64Maybe`, we could even put that function in a separate module and test it. `Data.Natural.Extras` or something.